src/lxc/commands.c

```diff
@@ -904,44 +904,35 @@ int lxc_cmd_add_state_client(const char *name, const char *lxcpath,
 	    },
 	};
 
-	/* Lock the whole lxc_cmd_add_state_client_callback() call to ensure
-	*  that lxc_set_state() doesn't cause us to miss a state.
-	 */
-	process_lock();
 	/* Check if already in requested state. */
 	state = lxc_getstate(name, lxcpath);
```

Comment:
...  but the state could change between lxc_getstate and later checks, no?

Reply:
I think the right thing to do is to remove this check altogether. Instead, we
check the state by looking at handler->state directly when we have taken the
lock on the state client list in lxc_add_state_client(), i.e. in the callback.